From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Katzen48 | Tobi <Katzen48@freenet.de>
Date: Mon, 16 Oct 2023 21:38:18 +0200
Subject: [PATCH] added entitybytag variable


diff --git a/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java b/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
index c4d1dbbd39ba0cdc9176ffa6d350d2aa50380211..58d8c070d649b204b362b674d3d9cf06a44a753b 100644
--- a/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
+++ b/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
@@ -1,6 +1,7 @@
 package io.papermc.paper.chunk.system.entity;
 
 import com.destroystokyo.paper.util.maplist.EntityList;
+import com.google.common.collect.ArrayListMultimap;
 import com.mojang.logging.LogUtils;
 import io.papermc.paper.util.CoordinateUtils;
 import io.papermc.paper.util.TickThread;
@@ -59,6 +60,9 @@ public final class EntityLookup implements LevelEntityGetter<Entity> {
     private final StampedLock entityByLock = new StampedLock();
     private final Int2ReferenceOpenHashMap<Entity> entityById = new Int2ReferenceOpenHashMap<>();
     private final Object2ReferenceOpenHashMap<UUID, Entity> entityByUUID = new Object2ReferenceOpenHashMap<>();
+    // ForkTest start
+    private final ArrayListMultimap<String, Entity> entitiesByTag = ArrayListMultimap.create();
+    // ForkTest end
     private final EntityList accessibleEntities = new EntityList();
 
     public EntityLookup(final ServerLevel world, final LevelCallback<Entity> worldCallback) {
